pkg/expression/builtin_control.go

```diff
@@ -117,37 +117,60 @@ func setDecimalFromArgs(evalType types.EvalType, resultFieldType *types.FieldTyp
 	}
 }
 
+// NonBinaryStr means the arg is a string but not binary string
+func getNonBinaryStrIdx(args []*types.FieldType) []int {
+	result := make([]int, 0)
+	for i, arg := range args {
+		if types.IsNonBinaryStr(arg) {
+			result = append(result, i)
+		}
+	}
+	return result
+}
+
+func hasBinaryStr(args []*types.FieldType) bool {
+	for _, arg := range args {
+		if types.IsBinaryStr(arg) {
+			return true
+		}
+	}
+	return false
+}
+
 func addCollateAndCharsetAndFlagFromArgs(ctx sessionctx.Context, funcName string, evalType types.EvalType, resultFieldType *types.FieldType, args ...Expression) error {
 	switch funcName {
-	case ast.If, ast.Ifnull, ast.WindowFuncLead, ast.WindowFuncLag:
-		if len(args) != 2 {
-			panic("unexpected length of args for if/ifnull/lead/lag")
+	case ast.If, ast.Ifnull, ast.WindowFuncLead, ast.WindowFuncLag, ast.Coalesce:
```

Comment:
maybe `coalesce` should be treated the same as `case`, so you don't need to change these codes